src/Tasks/DownloadFile.cs

```diff
@@ -77,36 +83,46 @@ public override bool Execute()
             }
 
             int retryAttemptCount = 0;
-            bool canRetry = false;
+            
+            CancellationToken cancellationToken = _cancellationTokenSource.Token;
 
-            do
+            while(true)
```

Comment:
Why did you change the type of loop? I personally would prefer to see a single while or do/while loop without break statements if possible, that is, putting `canRetry = false;` in instead of breaks.

Reply:
That is the source of the bugs.  There are many cases that _should_ break out of the loop, but failed to do so before, because it carried the value from the previous iteration of the loop unexpectedly.  There is no way to scope the variable to the body of loop, so you HAVE to carry the value between iterations, even though each iteration has nothing to do with the previous one.  By using explicit breaks, it's super clear when something means to break out of the loop.

I could change it back, and then just make sure to reset "canRetry" to false at the beginning of every loop... and then make sure it's set to true in every case where we want to retry... but that seems like a complicated and error prone way of saying "break" when we want to break.